### PR TITLE
Installer fine tuning

### PR DESCRIPTION
**Pull request by morph027**
- used [shellcheck](https://github.com/koalaman/shellcheck/) to lint the script
- added MariaDB repo to Ubuntu
- added logging function ```log()``` to get rid of duplicate code (adding results to logfile)

Tested on Debian / Ubuntu using Docker

### CODE REVIEW

**Comment by snipe** [comment]
Can you let me kmow when you think this is ready for review and merging? I don't want to merge and then have a bunch of follow up merges

**Comment by morph027** [comment]
Sure, I'll let you know! Updated the title for now...

By the way, thanks for this awesome project!

**Comment by morph027** [comment]
Ok, testing done, think it's ready to get reviewed...

**Comment by snipe** [comment]
@uberbrady can I get a sanity check on this before merge?

**Comment by therealjoshuad** [comment]
is suido a typo on line 19?

File: snipeit.sh [hunk 1]
```diff
@@ -288,9 +299,10 @@ case $distro in
 		curl -sS https://getcomposer.org/installer  | php
 		php composer.phar install --no-dev --prefer-source
 		perms
+		chown -R www-data:www-data "/var/www/$name"
```
Comment (uberbrady) [inline snipeit.sh]: Ah, I guess you *do* have the `chown -R` in other branches too! Good, that's at least consistent. Are we still sure it's a good idea?

File: snipeit.sh [hunk 2]
```diff
@@ -253,33 +260,37 @@ case $distro in
 		#composer install, set permissions, restart apache.
 
 		webdir=/var/www
+		echo -ne "\n* Adding MariaDB repoin the background... ${spin[0]}"
```
Comment (uberbrady) [inline snipeit.sh]: Should that be "repo in the background" ?

File: snipeit.sh [hunk 3]
```diff
@@ -299,35 +311,37 @@ case $distro in
...
+			echo "name = MariaDB"
+			echo "baseurl = http://yum.mariadb.org/10.0/centos6-amd64"
+			echo "gpgkey=https://yum.mariadb.org/RPM-GPG-KEY-MariaDB"
+			echo "gpgcheck=1"
+			echo "enable=1"
+		} >> "$mariadbRepo"
+
+		yum -y install wget epel-release >/dev/null 2>&1
+		wget -P "$tmp/" https://centos6.iuscommunity.org/ius-release.rpm >/dev/null 2>&1
+		rpm -Uvh "$tmp/ius-release*.rpm" >/dev/null 2>&1
```
Comment (uberbrady) [inline snipeit.sh]: No logs on these three?

File: snipeit.sh [hunk 4]
```diff
@@ -299,35 +311,37 @@ case $distro in
...
+				echo -n " ## installing $p ... "
+				yum -y install "$p" >/dev/null 2>&1
 				echo "";
 			fi
 		done;
 
 		echo -e "\n##  Downloading Snipe-IT from github and putting it in the web directory.";
 
-		wget -P $tmp/ https://github.com/snipe/snipe-it/archive/$file >> /var/log/snipeit-install.log 2>&1
+		wget -P "$tmp/" "https://github.com/snipe/snipe-it/archive/$file" >/dev/null 2>&1
```
Comment (uberbrady) [inline snipeit.sh]: more no log?

File: snipeit.sh [hunk 5]
```diff
@@ -391,10 +407,8 @@ case $distro in
 		curl -sS https://getcomposer.org/installer | php
 		php composer.phar install --no-dev --prefer-source
 
-		# Change permissions on directories
-		sudo chmod -R 755 $webdir/$name/storage
-		sudo chmod -R 755 $webdir/$name/public/uploads
-		sudo chown -R apache:apache $webdir/$name
+		perms
```
Comment (uberbrady) [inline snipeit.sh]: what's this?

File: snipeit.sh [hunk 6]
```diff
@@ -391,10 +407,8 @@ case $distro in
 		curl -sS https://getcomposer.org/installer | php
 		php composer.phar install --no-dev --prefer-source
 
-		# Change permissions on directories
-		sudo chmod -R 755 $webdir/$name/storage
-		sudo chmod -R 755 $webdir/$name/public/uploads
```
Comment (uberbrady) [inline snipeit.sh]: Why did we lose these chmods? Are we doing it somewhere else, or is it to be consistent with something else? Or that the `chown -R` will cover it?

File: snipeit.sh [hunk 7]
```diff
@@ -542,6 +556,6 @@ echo ""
 echo "* Cleaning up..."
 rm -f snipeit.sh
 rm -f install.sh
-rm -rf $tmp/
+rm -rf ${tmp:?}
```
Comment (uberbrady) [inline snipeit.sh]: That same bit about `:?` that I mentioned before - sure this is alright?

**Comment by morph027** [comment]
Ah, sure...weird. Looks like vim insert typo ;) And fixed...

File: snipeit.sh [hunk 8]
```diff
@@ -415,29 +429,29 @@ case $distro in
 
 		#Allow us to get the mysql engine
 		echo -e "\n##  Add IUS, epel-release and mariaDB repos.";
-		yum -y install wget epel-release >> /var/log/snipeit-install.log 2>&1
-		wget -P $tmp/ https://centos7.iuscommunity.org/ius-release.rpm >> /var/log/snipeit-install.log 2>&1
-		rpm -Uvh $tmp/ius-release*.rpm >> /var/log/snipeit-install.log 2>&1
+		yum -y install wget epel-release >/dev/null 2>&1
+		wget -P $tmp/ https://centos7.iuscommunity.org/ius-release.rpm >/dev/null 2>&1
+		rpm -Uvh $tmp/ius-release*.rpm >/dev/null 2>&1
```
Comment (uberbrady) [inline snipeit.sh]: No more logging in here?

File: snipeit.sh [hunk 9]
```diff
@@ -415,29 +429,29 @@ case $distro in
...
 		for p in $PACKAGES;do
-			if isinstalled $p;then
-				echo " ##" $p "Installed"
+			if isinstalled "$p"; then
+				echo " ## $p already installed"
 			else
-				echo -n " ##" $p "Installing... "
-				yum -y install $p >> /var/log/snipeit-install.log 2>&1
+				echo -n " ## installing $p ... "
+				yum -y install "$p" >/dev/null 2>&1
```
Comment (uberbrady) [inline snipeit.sh]: No logs?

File: snipeit.sh [hunk 10]
```diff
@@ -415,29 +429,29 @@ case $distro in
...
 			echo "";
 			fi
 		done;
 
 		echo -e "\n##  Downloading Snipe-IT from github and put it in the web directory.";
 
-		wget -P $tmp/ https://github.com/snipe/snipe-it/archive/$file >> /var/log/snipeit-install.log 2>&1
-		unzip -qo $tmp/$file -d $tmp/
-		cp -R $tmp/$fileName $webdir/$name
+		wget -P "$tmp/" "https://github.com/snipe/snipe-it/archive/$file" >/dev/null 2>&1
```
Comment (uberbrady) [inline snipeit.sh]: No log here?

File: snipeit.sh [hunk 11]
```diff
@@ -509,20 +525,18 @@ case $distro in
 		curl -sS https://getcomposer.org/installer | php
 		php composer.phar install --no-dev --prefer-source
 
-		sudo chmod -R 755 $webdir/$name/storage
-		sudo chmod -R 755 $webdir/$name/storage/private_uploads
-		sudo chmod -R 755 $webdir/$name/public/uploads
-		sudo chown -R apache:apache $webdir/$name
+		perms
```
Comment (uberbrady) [inline snipeit.sh]: Ah, there's that "perms" thing again. What's that?

File: snipeit.sh [hunk 12]
```diff
@@ -36,16 +35,16 @@ spin[1]="\\"
 spin[2]="|"
 spin[3]="/"
 
-rm -rf $tmp/
+rm -rf ${tmp:?}
```
Comment (morph027) [inline snipeit.sh]: Not really, this prevents deleting ```/``` if ```$tmp``` is unset.

https://github.com/koalaman/shellcheck/wiki/SC2115

The colon is bash builtin fallback for unset variables.

File: snipeit.sh [hunk 13]
```diff
@@ -56,24 +55,26 @@ progress () {
 }
 
 vhenvfile () {
-		sudo ls -al /etc/apache2/mods-enabled/rewrite.load >> /var/log/snipeit-install.log 2>&1
+		find /etc/apache2/mods-enabled -maxdepth 1 -name 'rewrite.load' >/dev/null 2>&1
```
Comment (morph027) [inline snipeit.sh]: https://github.com/koalaman/shellcheck/wiki/SC2012

File: snipeit.sh [hunk 14]
```diff
@@ -56,24 +55,26 @@ progress () {
...
+			echo "<Directory $webdir/$name/public>"
+			echo "        Require all granted"
+			echo "        AllowOverride All"
+			echo "   </Directory>"
+			echo "    DocumentRoot $webdir/$name/public"
+			echo "    ServerName $fqdn"
+			echo "        ErrorLog /var/log/apache2/snipeIT.error.log"
+			echo "        CustomLog /var/log/apache2/access.log combined"
+			echo "</VirtualHost>"
+		} >> $apachefile
```
Comment (morph027) [inline snipeit.sh]: :smile: Also this: https://github.com/koalaman/shellcheck/wiki/SC2129

File: snipeit.sh [hunk 15]
```diff
@@ -86,44 +87,46 @@ vhenvfile () {
...
+		true
+	else
+		false
+	fi
 }
 
 if [ -f /etc/lsb-release ]; then
-    distro="$(lsb_release -s -i )"
-    version="$(lsb_release -s -r)"
+	. /etc/lsb-release
```
Comment (morph027) [inline snipeit.sh]: I've tested using stock docker images for Debian and Ubuntu, the command did not exist in both. We should probably just install ```lsb-release``` package.

File: snipeit.sh [hunk 16]
```diff
@@ -86,44 +87,46 @@ vhenvfile () {
...
+	codename="$DISTRIB_CODENAME"
 elif [ -f /etc/os-release ]; then
-    distro="$(. /etc/os-release && echo $ID)"
-    version="$(. /etc/os-release && echo $VERSION_ID)"
-#Order is important here.  If /etc/os-release and /etc/centos-release exist, we're on centos 7.
-#If only /etc/centos-release exist, we're on centos6(or earlier).  Centos-release is less parsable,
-#so lets assume that it's version 6 (Plus, who would be doing a new install of anything on centos5 at this point..)
+	. /etc/os-release
+	distro="$ID"
+	version="$VERSION_ID"
```
Comment (morph027) [inline snipeit.sh]: Sounds reasonable, will revert this.

File: snipeit.sh [hunk 17]
```diff
@@ -216,34 +221,36 @@ case $distro in
...
 		# and just set the snipeit database user at the beginning
+		service mysql status >/dev/null || service mysql start
 		/usr/bin/mysql_secure_installation
 		echo -e "* Creating Mysql Database and User.\n##  Please Input your MySQL/MariaDB root password: "
 		mysql -u root -p < $dbsetup
 		cd $webdir/$name/
 		curl -sS https://getcomposer.org/installer | php
 		php composer.phar install --no-dev --prefer-source
 		perms
+		chown -R www-data:www-data "/var/www/$name"
```
Comment (morph027) [inline snipeit.sh]: Was inside https://github.com/snipe/snipe-it/blob/master/snipeit.sh#L94 before, but i've re-used perms() for all OS'es.

File: snipeit.sh (hunk 1 above)
Comment (morph027) [inline snipeit.sh]: Might work without. but not guaranteed ;)

File: snipeit.sh (hunk 3 above)
Comment (morph027) [inline snipeit.sh]: Added

File: snipeit.sh (hunk 4 above)
Comment (morph027) [inline snipeit.sh]: Added

File: snipeit.sh (hunk 6 above)
Comment (morph027) [inline snipeit.sh]: Added them to ```perms()```

File: snipeit.sh (hunk 5 above)
Comment (morph027) [inline snipeit.sh]: The ```perms()``` function which does the whole chmod things for all OS'es

File: snipeit.sh (hunk 8 above)
Comment (morph027) [inline snipeit.sh]: Added

File: snipeit.sh (hunk 9 above)
Comment (morph027) [inline snipeit.sh]: Added

File: snipeit.sh [hunk 18]
```diff
@@ -192,17 +195,19 @@ esac
 done
 
 #Snipe says we need a new 32bit key, so let's create one randomly and inject it into the file
-random32="$(echo `< /dev/urandom tr -dc _A-Za-z-0-9 | head -c32`)"
+random32="$(< /dev/urandom tr -dc _A-Za-z-0-9 2>&1 | head -c32)"
```
Comment (mtucker6784) [inline snipeit.sh]: I agree there, good call on reworking that line.

File: snipeit.sh (hunk 15 above)
Comment (n8felton) [inline snipeit.sh]: This caused a regression. See #3572.

**Pull request merged**